README.md
# dbsync
database sync tools.

## config
config.json file.

1. SrcDB source database config.
2. DesDB destcation database config.
3. DataSync data sync handle config.

eg : 

<p align="center"><img src="doc/config/eg.jpg" alt="eg Logo"></p>


## build
- make install
- make
- make release

## run
./bin/start.sh

## note
dbsync是数据库同步工具,使用go编写，主要特性有
1. 支持数据库种类多(oracle,postgresql,mysql,sqlserver等,其他添加中)
2. 灵活支持自定义SQL,支持多表和视图
3. 支持字段映射,可以配置字段的映射关系
4. 批量操作,性能较高

## benchmark
- 本地虚拟机 centos6 2核2G 
- PostgreSQL 10.6
- mysql 5.1.73

1. insert 117275 条记录 14S TPS 8376/s
2. update 117275 条记录 44S TPS 2665/s



CHANGELOG.md
## 2020.03.24

去掉 //_ "github.com/mattn/go-oci8"，会产生pkg的错误

暂时不考虑oracle的数据同步



db/db.go
//==================================
//  * Name：DataSync
//  * DateTime：2019/08/15
//  * File: db.go
//  * Note: db common handle .
//==================================

package db

import (
	"fmt"
	"strconv"
	"strings"

	"database/sql"

	_ "github.com/go-sql-driver/mysql"
	_ "github.com/lib/pq"
	_ "cloud.google.com/go"
	//_ "github.com/mattn/go-oci8"
)

func Setup() {

}

func (db *DBServer) Start() error {
	var err error

	dbtype := db.DBtype
	host := db.Host
	user := db.User
	passwd := db.Passwd
	dbname := db.DBName

	switch dbtype {
	case MYSQL:
		DSN := fmt.Sprintf("%s:%s@tcp(%s)/%s?charset=utf8", user, passwd, host, dbname)
		db.DB, err = sql.Open(MYSQL, DSN)
	case POSTGRESQL:
		DSN := fmt.Sprintf("postgres://%s:%s@%s/%s?sslmode=disable", user, passwd, host, dbname)
		db.DB, err = sql.Open(POSTGRESQL, DSN)
	case ORACLE:
		// [username/[password]@]host[:port][/instance_name]
		DSN := fmt.Sprintf("%s/%s@%s/orcl", user, passwd, host)
		db.DB, err = sql.Open("oci8", DSN)
	case SQLITE3:
		// ./testdb.db
		db.DB, err = sql.Open(SQLITE3, fmt.Sprintf("%s", dbname))
	case MSSQLSERVER:
		DSN := fmt.Sprintf("sqlserver://%s:%s@%s/instance?database=%s", user, passwd, host, dbname)
		db.DB, err = sql.Open(MSSQLSERVER, DSN)
	default:
		panic(fmt.Sprintf("InitDB error . dbtype %s is error .", dbtype))
	}

	if err != nil {
		panic(fmt.Sprintf("InitDB error . error message : %s", err))
	}
	db.DB.SetMaxIdleConns(DBConnMax / 2)
	db.DB.SetMaxOpenConns(DBConnMax)
	return nil
}

func (db *DBServer) Stop() error {
	if db.DB != nil {
		db.DB.Close()
	}
	return nil
}

func (db *DBServer) ReName(resultMap []map[string]string, reNameInfo ReNameMapInfo) []map[string]string {
	res := resultMap
	nameMap := reNameInfo.NameMap
	if nameMap == nil || len(nameMap) == 0 {
		return res
	}
	for i := 0; i < len(res); i++ {
		itemMap := res[i]
		for src, des := range nameMap {
			tmpStr := itemMap[src]
			itemMap[des] = tmpStr
			delete(itemMap, src)
		}
		res[i] = itemMap
	}
	// oracle handle . delete RN .
	if _, ok := res[0]["RN"]; ok {
		for i := 0; i < len(res); i++ {
			itemMap := res[i]
			delete(itemMap, "RN")
		}
	}
	return res
}

func dbQueryString(db *sql.DB, strsql string) ([]map[string]string, error) {
	rows, err := db.Query(strsql)
	fmt.Println("strsql : ", strsql)
	checkErr(err)
	cols, _ := rows.Columns()
	values := make([][]byte, len(cols))
	scans := make([]interface{}, len(cols))
	for i := range values {
		scans[i] = &values[i]
	}
	results := []map[string]string{}
	for rows.Next() {
		if err := rows.Scan(scans...); err != nil {
			return results, err
		}
		curRow := make(map[string]string)
		for k, v := range values {
			curRow[cols[k]] = string(v)
		}
		results = append(results, curRow)
	}
	rows.Close()
	return results, nil
}

func (db *DBServer) Query(queryInfo QueryInfo) []map[string]string {
	if db.DB != nil {
		sql := createQuerySql(queryInfo)
		results, err := dbQueryString(db.DB, sql)
		if err != nil {
			fmt.Println("Query err , sql : ", sql)
		}
		return results
	}
	return nil
}

func (db *DBServer) QuerySQL(sql string) []map[string]string {
	if db.DB != nil {
		results, err := dbQueryString(db.DB, sql)
		if err != nil {
			fmt.Println("Query err : ", err, "sql : ", sql)
		}
		return results
	}
	return nil
}

func (db *DBServer) Exec(execInfo ExecInfo) error {
	if db.DB != nil {
		sqlArr := createExecSql(db.DBtype, execInfo)
		for _, sql := range sqlArr {
			_, err := db.DB.Exec(sql)
			if err != nil {
				fmt.Println("Exec err : ", err, " sql : ", sql)
				return err
			}
		}
	}
	return nil
}

func (db *DBServer) ExecSQL(sql string) error {
	if db.DB != nil {
		_, err := db.DB.Exec(sql)
		if err != nil {
			fmt.Println("Exec err : ", err, " sql : ", sql)
			return err
		}
	}
	return nil
}

func (db *DBServer) TruncateTable(tablename string) error {
	if db.DB != nil {
		found := false
		for _, v := range db.TablesName {
			if v == tablename {
				found = true
			}
		}
		if found {
			sqlstr := "TRUNCATE " + tablename
			_, err := db.DB.Exec(sqlstr)
			checkErr(err)
		}
	}
	return nil
}

func createQuerySql(queryInfo QueryInfo) string {
	sql := ""

	keyStr := ""
	ConditionStr := ""
	// select key1,key2 from tablename where condition1,condition2 order by PK limit 1,100
	if (len(queryInfo.KeyArr) == 0) || (len(queryInfo.KeyArr) == 1 && queryInfo.KeyArr[0] == "*") {
		keyStr = "*"
	} else {
		keyStr = strCombine(queryInfo.KeyArr, ",")
	}
	if len(queryInfo.ConditionArr) > 0 {
		ConditionStr = "where "
		ConditionStr = ConditionStr + strCombine(queryInfo.ConditionArr, " AND ")
	}

	sql = fmt.Sprintf("select %s from %s %s order by %s", keyStr, queryInfo.TableName, ConditionStr, queryInfo.PK)
	return sql
}

func (db *DBServer) GetQueryTotalCount(sql string) uint {
	sqlLower := strings.ToLower(sql)
	pos := strings.Index(sqlLower, "from")
	if pos <= 0 {
		return 0
	}
	totalSql := "select count(*) as count " + sql[pos:]
	orderPos := strings.Index(sqlLower, "order")
	if orderPos > 0 {
		totalSql = "select count(*) as count " + sql[pos:orderPos]
	}
	if totalSql != "" {
		res := db.QuerySQL(totalSql)
		if len(res) > 0 {
			strCount, ok := res[0]["count"]
			if ok == false {
				// for oracle handle.
				strCount = res[0]["COUNT"]
			}
			count, _ := strconv.Atoi(strCount)
			return uint(count)
		}
	}
	return 0
}

func (db *DBServer) GetMaxValue(field string) string {
	sql := ""
	if db.DBtype == ORACLE {
		sql = fmt.Sprintf(" select %s from %s where rownum <= 1 order by %s desc ", field, db.TablesName, field)
	}
	if db.DBtype == MYSQL || db.DBtype == POSTGRESQL {
		sql = fmt.Sprintf(" select %s from %s order by %s desc limit 1 ", field, db.TablesName, field)
	}
	res := db.QuerySQL(sql)
	if len(res) > 0 {
		strCount, ok := res[0][field]
		if ok == false {
			fmt.Println("GetMaxValue error , field : ", field, " not in ", db.TablesName)
		}
		return strCount
	}
	return ""
}

func (db *DBServer) QueryPage(sql string, start int, num int) string {
	if db.DBtype == ORACLE {
		tmpFmt := `
		SELECT *
		FROM (SELECT a.*, ROWNUM rn
				FROM (%s) a
				WHERE ROWNUM <= %d)
		WHERE rn > %d
		`
		return fmt.Sprintf(tmpFmt, sql, (num + start), start)
	}
	if db.DBtype == MYSQL {
		return sql + fmt.Sprintf(" limit %d,%d ", start, num)
	}
	if db.DBtype == POSTGRESQL {
		return sql + fmt.Sprintf(" limit %d offset %d ", num, start)
	}
	if db.DBtype == MSSQLSERVER {
		/*
			tmpFmt := `
			SELECT  *
			FROM    ( SELECT TOP PageSize
								*
					FROM      ( SELECT TOP [PageIndex*PageSize]
											*
								FROM      (%s)
								ORDER BY  id ASC
								) AS b
					ORDER BY  id DESC
					) AS c
			ORDER BY id ASC;
			`
			return fmt.Sprintf(tmpFmt, sql, (num + start), start)
		*/
	}
	return sql
}

func (db *DBServer) GetTableColumns(tableName string) []map[string]string {
	sql := ""
	if db.DBtype == ORACLE {
		strFmt := "select COLUMN_NAME as NAME,DATA_TYPE as TYPE from user_tab_columns where Table_Name='%s' "
		sql = fmt.Sprintf(strFmt, tableName)
	}
	if db.DBtype == POSTGRESQL {
		strFmt := ` SELECT a.attname AS name,t.typname AS type FROM pg_class c,pg_attribute a,pg_type t
		WHERE c.relname = '%s' and a.attnum > 0 and a.attrelid = c.oid and a.atttypid = t.oid
		ORDER BY a.attnum `
		sql = fmt.Sprintf(strFmt, tableName)
	}
	if db.DBtype == MYSQL {
		strFmt := "SELECT COLUMN_NAME as name,DATA_TYPE as type FROM information_schema.columns WHERE table_name='%s'"
		sql = fmt.Sprintf(strFmt, tableName)
	}
	if db.DBtype == MSSQLSERVER {
		strFmt := "SELECT COLUMN_NAME as name,DATA_TYPE as type FROM information_schema.columns WHERE table_name='%s'"
		sql = fmt.Sprintf(strFmt, tableName)
	}
	results, err := dbQueryString(db.DB, sql)
	if err != nil {
		fmt.Println("Query err : ", err, " sql : ", sql)
	}
	return results
}

func createQueryTotalCountSql(queryInfo QueryInfo) string {
	sql := ""
	keyStr := ""
	ConditionStr := ""
	keyStr = " count(*) as count "
	if len(queryInfo.ConditionArr) > 0 {
		ConditionStr = "where "
		ConditionStr = ConditionStr + strCombine(queryInfo.ConditionArr, " AND ")
	}
	sql = fmt.Sprintf("select %s from %s %s order by %s", keyStr, queryInfo.TableName, ConditionStr, queryInfo.PK)
	return sql
}

func createExecSql(dbtype string, execInfo ExecInfo) []string {
	sqlArr := []string{}
	if execInfo.Handle == INSERT {
		sqlArr = insertSql(dbtype, execInfo)
	}
	if execInfo.Handle == UPDATE {
		sqlArr = updateSql(dbtype, execInfo)
	}
	return sqlArr
}

func insertSql(dbtype string, execInfo ExecInfo) []string {
	sql := ""
	sqlArr := []string{}

	results := execInfo.Content
	tablename := execInfo.TableName
	ColumnArr := []string{}
	keyStr, valueStr := "", ""
	for k, _ := range results[0] {
		keyStr += k + ","
		ColumnArr = append(ColumnArr, k)
	}
	keyStr = keyStr[:len(keyStr)-1]

	if dbtype == MYSQL || dbtype == POSTGRESQL || dbtype == SQLITE3 {
		for i := 0; i < len(results); i += MAXSQLCOUNT {
			valueStr = "VALUES"
			endflag := false
			for count := 0; count < MAXSQLCOUNT && !endflag; count++ {
				RowTmpValueStr := "("
				for j := 0; j < len(ColumnArr); j++ {
					if (i + count) >= len(results) {
						endflag = true
						break
					}
					TmpValueStr := results[i+count][ColumnArr[j]]
					if TmpValueStr == "<nil>" {
						TmpValueStr = ""
					}
					if strings.Contains(TmpValueStr, "'") {
						TmpValueStr = strings.Replace(TmpValueStr, "'", "''", -1)
					}

					//edit by raoweili, 2020.03.24, 非字符串，如果是空置，则设置成null
					if(execInfo.ColumnMap[ColumnArr[j]] != "varchar" && TmpValueStr == "") {
						RowTmpValueStr += "null" + ","
					} else {
						RowTmpValueStr += "'" + TmpValueStr + "'" + ","
					}

				}
				if endflag {
					RowTmpValueStr = ""
				} else {
					RowTmpValueStr = RowTmpValueStr[:len(RowTmpValueStr)-1]
					RowTmpValueStr += "),"
					valueStr += RowTmpValueStr
				}
			}
			valueStr = valueStr[:len(valueStr)-1]
			sqlstr := "INSERT INTO " + tablename + " (" + keyStr + " ) " + valueStr
			sqlArr = append(sqlArr, sqlstr)
		}
		return sqlArr
	}
	if dbtype == ORACLE {
		beginStr := "BEGIN \r\n"
		endStr := "\r\n END;"

		insertAllStr := ""
		insertStr := fmt.Sprintf(" INSERT INTO %s (%s) ", tablename, keyStr)
		for i := 0; i < len(results); i += MAXSQLCOUNT {
			valueStrFmt := " VALUES (%s) ; "
			valueStr := ""
			endflag := false
			for count := 0; count < MAXSQLCOUNT && !endflag; count++ {
				RowTmpValueStr := ""
				for j := 0; j < len(ColumnArr); j++ {
					if (i + count) >= len(results) {
						endflag = true
						break
					}
					TmpValueStr := results[i+count][ColumnArr[j]]
					if strings.Contains(TmpValueStr, "'") {
						TmpValueStr = strings.Replace(TmpValueStr, "'", "''", -1)
					}
					RowTmpValueStr += "'" + TmpValueStr + "'" + ","
				}
				if endflag {
					RowTmpValueStr = ""
				} else {
					RowTmpValueStr = RowTmpValueStr[:len(RowTmpValueStr)-1]
					valueStr = fmt.Sprintf(valueStrFmt, RowTmpValueStr)
					insertAllStr = insertAllStr + insertStr + valueStr
				}
			}
		}
		sql = beginStr + insertAllStr + endStr
		sqlArr = append(sqlArr, sql)
		return sqlArr
	}
	return sqlArr
}

func updateSql(dbtype string, execInfo ExecInfo) []string {

	// update table set key=value where id=1
	sqlFmt := " update %s set %s where %s='%s' ; "
	content := execInfo.Content
	sqlArr := []string{}
	count := len(execInfo.Content)

	for i := 0; i < count; i += MAXSQLCOUNT {
		sqlTotal := ""
		sql := ""
		endflag := false
		for j := 0; j < MAXSQLCOUNT && !endflag; j++ {
			strKey := ""
			PKValue := ""
			if (i + j) >= count {
				endflag = true
				break
			}
			for k, v := range content[i+j] {
				if k != execInfo.PK {
					strKey = strKey + k + "='" + v + "',"
				} else {
					PKValue = v
				}
			}
			strKey = strKey[:len(strKey)-1]
			sql = fmt.Sprintf(sqlFmt, execInfo.TableName, strKey, execInfo.PK, PKValue)
			sqlTotal = sqlTotal + sql
			sqlArr = append(sqlArr, sqlTotal)
		}
	}
	return sqlArr
}

func strCombine(arr []string, comb string) string {
	res := ""
	for k, v := range arr {
		if k == len(arr)-1 {
			res += v
		} else {
			res += v + comb
		}
	}
	return res
}


handle/handle.go
//==================================
//  * Name：DataSync
//  * DateTime：2019/08/16
//  * File: handle.go
//  * Note: Business processing.
//==================================

package handle

import (
	"crypto/md5"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/tidwall/gjson"

	"go-dbsync/config"
	"go-dbsync/db"
	"go-dbsync/pkg/cache"
	"go-dbsync/pkg/logging"
	gcache "github.com/patrickmn/go-cache"
)

var SrcDBPtr *db.DBServer
var DesDBPtr *db.DBServer

func Setup() {
	logging.Info(fmt.Sprintf("DataSync handle Setup , read config ."))
	//filepath := system.GetCurrentDirectory() + "/config.json"
	filepath := "config.json"
	config.ParseConfig(filepath)
	//cache init.
	cache.CacheInit()
}

func Work() {
	logging.Info(fmt.Sprintf("DataSync handle work."))
	db.Setup()
	//Get config.
	SrcDB := db.DBServer{}

	SrcDB.DBtype = gjson.Get(config.ConfigStringJSON, "SrcDB.type").String()
	SrcDB.User = gjson.Get(config.ConfigStringJSON, "SrcDB.User").String()
	SrcDB.Passwd = gjson.Get(config.ConfigStringJSON, "SrcDB.Passwd").String()
	SrcDB.Host = gjson.Get(config.ConfigStringJSON, "SrcDB.Host").String()
	SrcDB.DBName = gjson.Get(config.ConfigStringJSON, "SrcDB.DBName").String()
	err := SrcDB.Start()
	if err != nil {
		logging.Info(fmt.Sprintf("SrcDB Link Fail,Please check SrcDB config."))
		panic("SrcDB Link Fail,Please check SrcDB config.")
	}
	SrcDBPtr = &SrcDB
	DesDB := db.DBServer{}
	DesDB.DBtype = gjson.Get(config.ConfigStringJSON, "DesDB.type").String()
	DesDB.User = gjson.Get(config.ConfigStringJSON, "DesDB.User").String()
	DesDB.Passwd = gjson.Get(config.ConfigStringJSON, "DesDB.Passwd").String()
	DesDB.Host = gjson.Get(config.ConfigStringJSON, "DesDB.Host").String()
	DesDB.DBName = gjson.Get(config.ConfigStringJSON, "DesDB.DBName").String()
	err = DesDB.Start()
	if err != nil {
		logging.Info(fmt.Sprintf("DesDB Link Fail,Please check DesDB config."))
		panic("DesDB Link Fail,Please check DesDB config.")
	}
	DesDBPtr = &DesDB

	jobMap := gjson.Get(config.ConfigStringJSON, "DataSync.0.job").Map()
	sql := jobMap["srcSql"].String()
	//get FieldsMap config.
	reNameConfig := jobMap["FieldsMap"].Map()
	reNameMapInfo := db.ReNameMapInfo{}
	reNameMapInfo.SrcTableName = jobMap["srcTable"].String()
	reNameMapInfo.DesTableName = jobMap["desTable"].String()
	NameMap := make(map[string]string)
	for k, v := range reNameConfig {
		NameMap[k] = v.String()
	}
	reNameMapInfo.NameMap = NameMap
	// get srcDB tableColumns
	tableColumnsResult := SrcDB.GetTableColumns(jobMap["srcTable"].String())
	tableColumnsMap := make(map[string]string)
	for i := 0; i < len(tableColumnsResult); i++ {
		key := ""
		val := ""
		if SrcDB.DBtype == db.ORACLE {
			key = tableColumnsResult[i]["NAME"]
			val = tableColumnsResult[i]["TYPE"]
		}
		if SrcDB.DBtype == db.POSTGRESQL || SrcDB.DBtype == db.MYSQL {
			key = tableColumnsResult[i]["name"]
			val = tableColumnsResult[i]["type"]
		}
		if res, ok := NameMap[key]; ok {
			tableColumnsMap[res] = val
		} else {
			tableColumnsMap[key] = val
		}
	}
	strColumns, _ := json.Marshal(tableColumnsMap)

	logging.Info(fmt.Sprintf("tableColumnsMap : %s", string(strColumns)))
	//get Des Table to redis.
	srcTableName := jobMap["srcTable"].String()
	desTableName := jobMap["desTable"].String()
	sql = "select count(*) as count from " + desTableName
	desTotalCount := DesDB.GetQueryTotalCount(sql)

	for i := 0; i < int(desTotalCount); i += db.MAXPAGECOUNT {
		sql := "select * from " + desTableName
		sql = DesDBPtr.QueryPage(sql, i, db.MAXPAGECOUNT)
		desTableResult := DesDBPtr.QuerySQL(sql)
		CacheToLocal(desTableName, desTableResult)
	}

	// full or incr handle.
	syncType := jobMap["syncType"].String()
	if syncType == FULL {
		//support page query. get total count .
		sql = jobMap["srcSql"].String()
		totalCount := SrcDB.GetQueryTotalCount(sql)
		logging.Info(fmt.Sprintf("FULL handle , Source tableName : %s , totalCount : %d ", srcTableName, totalCount))
		logging.Info(fmt.Sprintf("FULL handle , Destination tableName : %s , totalCount : %d", desTableName, desTotalCount))
		if totalCount > db.MAXPAGECOUNT {
			start := 0
			num := db.MAXPAGECOUNT
			for start = 0; start < int(totalCount); start += db.MAXPAGECOUNT {
				logging.Info(fmt.Sprintf("Paging query . start : %d,num : %d. handled : %d .", start, num, start))
				sqlQueryPage := SrcDB.QueryPage(sql, start, num)
				//
				srcTableResult := SrcDB.QuerySQL(sqlQueryPage)
				renameResult := SrcDB.ReName(srcTableResult, reNameMapInfo)
				//renameResultstr, _ := json.Marshal(renameResult)
				//logging.Info("ReNameResult : ", string(renameResultstr))
				insertExec, updateExec := CompareWithCache(tableColumnsMap, renameResult)
				InsertAndUpdate(insertExec, updateExec)
			}
		} else {
			//
			srcTableResult := SrcDB.QuerySQL(sql)
			renameResult := SrcDB.ReName(srcTableResult, reNameMapInfo)
			//renameResultstr, _ := json.Marshal(renameResult)
			//logging.Info("ReNameResult : ", string(renameResultstr))
			insertExec, updateExec := CompareWithCache(tableColumnsMap, renameResult)
			InsertAndUpdate(insertExec, updateExec)
		}
	}
	if syncType == INCR {
		desPK := jobMap["desTablePK"].String()
		desPKMax := DesDBPtr.GetMaxValue(desPK)
		sql = jobMap["srcSql"].String()
		if desPKMax != "" {
			condition := fmt.Sprintf(" %s >= %s ", desPK, desPKMax)
			sql = sqlAddCondition(srcTableName, sql, condition)
		}
		totalCount := SrcDB.GetQueryTotalCount(sql)
		logging.Info(fmt.Sprintf("INCR handle , Source tableName : %s , totalCount : %d ", srcTableName, totalCount))
		if totalCount > db.MAXPAGECOUNT {
			start := 0
			num := db.MAXPAGECOUNT
			for start = 0; start < int(totalCount); start += db.MAXPAGECOUNT {
				logging.Info(fmt.Sprintf("Paging query . start : %d,num : %d. handled : %d .", start, num, start))
				sqlQueryPage := SrcDB.QueryPage(sql, start, num)
				//
				srcTableResult := SrcDB.QuerySQL(sqlQueryPage)
				renameResult := SrcDB.ReName(srcTableResult, reNameMapInfo)
				//renameResultstr, _ := json.Marshal(renameResult)
				//logging.Info("ReNameResult : ", string(renameResultstr))
				insertExec, updateExec := CompareWithCache(tableColumnsMap, renameResult)
				InsertAndUpdate(insertExec, updateExec)
			}
		} else {
			srcTableResult := SrcDB.QuerySQL(sql)
			renameResult := SrcDB.ReName(srcTableResult, reNameMapInfo)
			//renameResultstr, _ := json.Marshal(renameResult)
			//logging.Info("ReNameResult : ", string(renameResultstr))
			insertExec, updateExec := CompareWithCache(tableColumnsMap, renameResult)
			InsertAndUpdate(insertExec, updateExec)
		}
	}
	time.Sleep(1 * time.Second)
}

func CacheToLocal(tableName string, desTableResult []map[string]string) {
	jobMap := gjson.Get(config.ConfigStringJSON, "DataSync.0.job").Map()
	tablePK := jobMap["desTablePK"].String()
	for _, v := range desTableResult {
		PK := v[tablePK]
		strVal, _ := json.Marshal(v)
		key := tableName + "_" + PK
		cache.GCache.Set(key, string(strVal), gcache.NoExpiration)
		//logging.Info(fmt.Sprintf("Set to Cache key : %s, val : %s", key, string(strVal)))
	}
}

func GetLocalCache(tableName string, PK string) string {
	key := tableName + "_" + PK
	if val, found := cache.GCache.Get(key); found {
		return val.(string)
	}
	return ""
}

func CompareWithCache(tableColumnsMap map[string]string, renameResult []map[string]string) (db.ExecInfo, db.ExecInfo) {
	// Compare.
	jobMap := gjson.Get(config.ConfigStringJSON, "DataSync.0.job").Map()
	tableName := jobMap["desTable"].String()
	tablePK := jobMap["desTablePK"].String()

	keysArr := []string{}
	for k, _ := range renameResult[0] {
		keysArr = append(keysArr, k)
	}
	//logging.Info(keysArr)
	insertExec := db.ExecInfo{DesDBPtr.DBName, tableName, db.INSERT, tablePK, []map[string]string{}, tableColumnsMap}
	updateExec := db.ExecInfo{DesDBPtr.DBName, tableName, db.UPDATE, tablePK, []map[string]string{}, tableColumnsMap}
	for _, v := range renameResult {
		handle := ""
		srcQueryStr := ""
		cacheQueryStr := ""
		srcQueryMD5 := ""
		cacheQueryMD5 := ""

		for i := 0; i < len(keysArr); i++ {
			if !isCompare(tableColumnsMap[keysArr[i]]) {
				continue
			}
			str := v[keysArr[i]] + ";"
			if res, resstr := TypeConv(tableColumnsMap[keysArr[i]], v[keysArr[i]]); res {
				str = resstr + ";"
			}
			srcQueryStr = srcQueryStr + str
			//
			valJSONStr := GetLocalCache(tableName, v[tablePK])
			if valJSONStr == "" {
				// key is not exist. Inset hande.
				handle = db.INSERT
				break
			}
			valMap := make(map[string]interface{})
			err := json.Unmarshal([]byte(valJSONStr), &valMap)
			if err != nil {
				handle = db.INSERT
				break
			}
			val := valMap[keysArr[i]].(string)
			str = val + ";"
			cacheQueryStr = cacheQueryStr + str
		}
		//strShow, _ := json.Marshal(v)
		//logging.Info(fmt.Sprintf("need handle : %s , value :  %s, srcQueryStr : %s , cacheQueryStr : %s", handle, strShow, srcQueryStr, cacheQueryStr))
		if handle == "" {
			srcQueryMD5 = fmt.Sprintf("%x", md5.Sum([]byte(srcQueryStr)))
			cacheQueryMD5 = fmt.Sprintf("%x", md5.Sum([]byte(cacheQueryStr)))
			//logging.Info(fmt.Sprintf("need handle : %s , value :  %s, srcQueryStr : %s , cacheQueryStr : %s", handle, strShow, srcQueryStr, cacheQueryStr))
			//logging.Info(fmt.Sprintf("srcQueryMD5 : %s , cacheQueryMD5 : %s", srcQueryMD5, cacheQueryMD5))
			if srcQueryMD5 != cacheQueryMD5 {
				// update handle.
				handle = db.UPDATE
			}
		}

		if handle == db.INSERT {
			insertExec.Content = append(insertExec.Content, v)
		}
		if handle == db.UPDATE {
			updateExec.Content = append(updateExec.Content, v)
		}
	}
	return insertExec, updateExec
}

func InsertAndUpdate(insertExec db.ExecInfo, updateExec db.ExecInfo) {
	//insert handle.
	if len(insertExec.Content) > 0 {
		logging.Info(fmt.Sprintf("insertExec Count : %d ", len(insertExec.Content)))
		err := DesDBPtr.Exec(insertExec)
		if err != nil {
			logging.Info(err)
			os.Exit(0)
		}
		CacheToLocal(insertExec.TableName, insertExec.Content)
	}
	//update handle.
	if len(updateExec.Content) > 0 {
		logging.Info(fmt.Sprintf("updateExec Count : %d ", len(updateExec.Content)))
		err := DesDBPtr.Exec(updateExec)
		if err != nil {
			logging.Info(err)
		}
		CacheToLocal(insertExec.TableName, insertExec.Content)
	}
}

func sqlAddCondition(tableName string, sql string, condition string) string {
	tableNameTmp := strings.ToLower(tableName)
	sqlTmp := strings.ToLower(sql)

	whereFlag := strings.Contains(sqlTmp, "where")
	if whereFlag {
		pos := strings.Index(sqlTmp, "where")
		pos += 5
		sqlHead := sql[:pos]
		sqlTail := sql[pos:]
		return sqlHead + " " + condition + " and " + sqlTail
	} else {
		pos := strings.Index(sqlTmp, tableNameTmp)
		pos += len(tableName)
		sqlHead := sql[:pos]
		sqlTail := sql[pos:]
		return sqlHead + " where " + condition + sqlTail
	}
}

func isCompare(nameType string) bool {
	if nameType == "DATE" || nameType == "timestamp" {
		return false
	}
	return true
}

func TypeConv(nameType string, value string) (bool, string) {
	valueStr := strings.ToLower(value)
	//bool type handle, conv t -> true,f -> false.
	if (nameType == "VARCHAR2") && (valueStr == "t" || valueStr == "f") {
		if valueStr == "t" {
			return true, "true"
		}
		if valueStr == "f" {
			return true, "false"
		}
	}
	//DATE
	if nameType == "DATE" {

	}
	return false, ""
}


db/config.go
//==================================
//  * Name：DataSync
//  * DateTime：2019/08/15
//  * File: db.go
//  * Note: db common config .
//==================================

package db

import (
	"log"

	"database/sql"
)

const (
	MYSQL       = "mysql"
	POSTGRESQL  = "postgres"
	ORACLE      = "oracle"
	SQLITE3     = "sqlite3"
	MSSQLSERVER = "sqlserver"
)

const (
	INSERT = "insert"
	UPDATE = "update"
)

type DBServer struct {
	DBtype      string
	DBDriveName string
	User        string
	Passwd      string
	Host        string
	DBName      string
	DB          *sql.DB
	TablesName  []string
	TableKeys   map[string][]string
}

type QueryInfo struct {
	DBName       string
	TableName    string
	PK           string
	KeyArr       []string
	ConditionArr []string
	Content      map[string]string
}

type ExecInfo struct {
	DBName    string
	TableName string
	Handle    string
	PK        string
	Content   []map[string]string
	ColumnMap map[string]string
}

type ReNameMapInfo struct {
	SrcTableName string
	DesTableName string
	NameMap      map[string]string
}

const DBConnMax = 8
const MAXSQLCOUNT = 1000
const MAXPAGECOUNT = 20000

func init() {

}

func checkErr(err error) {
	if err != nil {
		log.Println(err)
	}
}

func checkErrPanic(err error) {
	if err != nil {
		log.Println(err)
		panic(err)
	}
}
